Implement serde for file uploader widget

AnOctopus opened the pull request:
The file uploader is weird, because the values it communicates with the
frontend are not directly translatable into regular python values. They
must be used to look up the file information from the ReportContext.
This unfortunately results in deserialization requiring the widget id,
which no other widget requires, so they all need to take and ignore this
second parameter.

Several tests for file uploader worked by patching `register_widget`,
which now returns the deserialized values instead of leaving some work
to the widget function. These tests will need to be rewritten to work
for this, but I wanted to get the changes up for review first, since it
might be involved.

AnOctopus committed:
Implement serde for file uploader widget

The file uploader is weird, because the values it communicates with the
frontend are not directly translatable into regular python values. They
must be used to look up the file information from the ReportContext.
This unfortunately results in deserialization requiring the widget id,
which no other widget requires, so they all need to take and ignore this
second parameter.

Several tests for file uploader worked by patching `register_widget`,
which now returns the deserialized values instead of leaving some work
to the widget function. These tests will need to be rewritten to work
for this, but I wanted to get the changes up for review first, since it
might be involved.

vdonato commented on lib/streamlit/elements/file_uploader.py:
```diff
@@ -147,11 +147,33 @@ def file_uploader(
 
         # TODO: Fix this (de)serializer so that st.session_state[key] can be
```
nit: can remove this TODO now

vdonato reviewed (approved):
Aside from the file_uploader test needing to be reworked, LGTM

One thing we can do to easily get some more test coverage on this before then (we already have some from the `st_file_uploader` e2e test) is to modify that test to also print the contents of the file uploader return value by fetching it from `st.session_state`

AnOctopus committed:
Remove unneeded file uploader comments

TODO done!

AnOctopus committed:
Add file uploader e2e state checks

To make up for temporary removal of some unit tests. Hopefully they
work, because I can't get even the unchanged parts to pass locally.

AnOctopus committed:
Wrap e2e state references in check for running st

AnOctopus committed:
Merge branch 'develop' into file-uploader-serde

AnOctopus committed:
Merge branch 'develop' into file-uploader-serde
